packages/backend/src/modules/tvl/api/services/data/DataStatusService.ts

```diff
@@ -0,0 +1,229 @@
+  | EscrowEntry
+  | CoingeckoPriceConfigEntry
+const MAX_CONFIGURATIONS_LENGTH_FOR_QUERY = 100
+
+export class DataStatusService {
+  constructor(private readonly db: Database) {}
+
+  async getAmountsStatus(
+    entries: (AmountConfigEntry & { configId: string })[],
+    targetTimestamp: UnixTime,
+  ) {
+    const configurations = await this.getConfigurationsStatus(
+      entries,
+      targetTimestamp,
+    )
+
+    const preminted = await this.getPremintedStatus(entries, targetTimestamp)
+
+    const circulating = await this.getCirculatingSupplyStatus(
+      entries,
+      targetTimestamp,
+    )
+
+    return {
+      lagging: [
+        ...configurations.lagging,
+        ...preminted.lagging,
+        ...circulating.lagging,
+      ],
+      excluded: new Set([
+        ...configurations.excluded,
+        ...preminted.excluded,
+        ...circulating.excluded,
+      ]),
+    }
+  }
+
+  async getPremintedStatus(
+    entries: (AmountConfigEntry & { configId: string })[],
+    targetTimestamp: UnixTime,
+  ) {
+    return this.getStatus(
+      entries.filter(
+        (c): c is PremintedEntry & { configId: string } =>
+          c.type === 'preminted',
+      ),
+      targetTimestamp,
+      (c) => PremintedIndexer.getId(c.chain, c.address),
+      (indexer, entries) =>
+        entries.find(
+          (cc) =>
+            cc.chain === indexer.indexerId.split('::')[1].split('_')[0] &&
+            cc.address === indexer.indexerId.split('::')[1].split('_')[1],
+        ),
+    )
+  }
+
+  async getCirculatingSupplyStatus(
+    entries: (AmountConfigEntry & { configId: string })[],
+    targetTimestamp: UnixTime,
+  ) {
+    return this.getStatus(
+      entries.filter(
+        (c): c is CirculatingSupplyEntry & { configId: string } =>
+          c.type === 'circulatingSupply',
+      ),
+      targetTimestamp,
+      (c) => CirculatingSupplyIndexer.getId(c.coingeckoId),
+      (indexer, entries) =>
+        entries.find(
+          (cc) =>
+            cc.coingeckoId.toString() === indexer.indexerId.split('::')[1],
+        ),
+    )
+  }
+
+  private async getStatus<T extends AmountConfigEntry & { configId: string }>(
+    entries: T[],
+    targetTimestamp: UnixTime,
+    getIndexerId: (entry: T) => string,
+    findConfig: (indexer: IndexerStateRecord, entries: T[]) => T | undefined,
+  ) {
+    const processed = new Set<string>()
+    const lagging: { id: string; latestTimestamp: UnixTime }[] = []
+    const excluded = new Set<string>()
+
+    const indexerState = await this.db.indexerState.getByIndexerIds(
+      entries.map(getIndexerId),
+    )
+
+    for (const indexer of indexerState) {
+      const config = findConfig(indexer, entries)
+      assert(config, `Config should be defined for ${indexer.indexerId}`)
+      processed.add(config.configId)
+
+      const syncStatus = new UnixTime(indexer.safeHeight)
+      if (syncStatus.gte(targetTimestamp)) {
+        continue
+      }
+
+      // TODO: what about max height?
```

Comment:
Should it be resolved now? If not do we have a ticket fot this?

Reply:
I will create a ticket